discovery_engine_core/core/src/engine.rs

```diff
@@ -324,35 +324,60 @@ where
-                let id = stack.id();
-                articles
-                    .into_iter()
-                    .map(|article| {
-                        let title = article.title.as_str();
-                        let embedding = self.ranker.compute_smbert(title).map_err(Error::Ranker)?;
-                        document_from_article(article, id, embedding).map_err(Error::Document)
-                    })
-                    .collect::<Result<Vec<_>, _>>()
-            }) {
-                Ok(documents) => {
-                    if let Err(error) = stack.update(&documents, &mut self.ranker) {
-                        errors.push(Error::StackOpFailed(error));
-                    } else {
-                        stack.data.retain_top(self.core_config.keep_top);
-                    }
+        let max_errors = stacks.len();
+        for stack in stacks {
+            let articles = match stack.new_items(&key_phrases).await {
+                Ok(articles) => articles,
+                Err(error) => {
+                    let error = Error::StackOpFailed(error);
+                    error!("{}", error);
+                    errors.push(error);
+                    continue;
+                }
+            };
+
+            let articles = match stack.filter_articles(history, articles) {
+                Ok(articles) => articles,
+                Err(error) => {
+                    let error = Error::StackOpFailed(error);
+                    error!("{}", error);
+                    errors.push(error);
+                    continue;
                 }
-                Err(error) => errors.push(error),
+            };
+
+            let id = stack.id();
+            let (documents, article_errors) = articles
+                .into_iter()
+                .map(|article| {
+                    let title = article.title.as_str();
+                    let embedding = self.ranker.compute_smbert(title).map_err(|error| {
+                        let error = Error::Ranker(error);
+                        error!("{}", error);
+                        error
+                    })?;
+                    document_from_article(article, id, embedding).map_err(|error| {
+                        let error = Error::Document(error);
+                        error!("{}", error);
+                        error
+                    })
+                })
+                .partition_result::<Vec<_>, Vec<_>, _, _>();
+            if documents.is_empty() {
+                errors.push(Error::Errors(article_errors));
+                continue;
+            }
+
+            if let Err(error) = stack.update(&documents, &mut self.ranker) {
+                let error = Error::StackOpFailed(error);
+                error!("{}", error);
+                errors.push(error);
+            } else {
+                stack.data.retain_top(self.core_config.keep_top);
             }
         }
 
-        if errors.is_empty() {
+        if errors.len() < max_errors {
```

Comment:
Can't we inline `stack.len()` here. or assign `max_errors` before this if?

Reply:
the `stacks` is moved in l.328, that's why `max_errors` is already assigned in l.327.

instead of moving `stacks` we could also mutably iterate over it, what do you prefer?

Reply:
If is adding only `&mut` in front I would do it and keep the information near for the error. If it's more work this is fine.